### ISSUE TITLE
LiteNetwork VNext proposals

### ISSUE THREAD

**Issue by Eastrall** [Eastrall/LiteNetwork#47]
Based on discussion and feedback on PR #46, I believe this can lead us to improve the library and make it even more easy to use.

# Protocol

* Remove the `ILitePacketStream` interface and all its dependencies

By removing this component, we also remove the dependency with the `ILitePacketStream` in the User and Client's `HandleMessageAsync(ILitePacketStream)` method and pass instead the real `byte[]`. This will allow the user to manipulate the raw buffer according to its needs:
```csharp
public class User : LiteServerUser
{
    public override Task HandleMessageAsync(byte[] rawBuffer)
    {
        using var reader = new BinaryReader(rawBuffer);

        int packetHeader = reader.ReadInt32(); // for example
    }
}
``` 

* Add `Send(byte[])` and `Send(Stream)` methods to `LiteClient` and `LiteServerUser`.
  * The `Send(byte[])` sends a raw packet to the remote end point.
  * The `Send(Stream)` method sends the buffer contained in the given stream to the remote end point.

By providing these two options, the user is free to choose which technique to use to send its packet. It can use the native .NET `BinaryWriter` class or any other way to build its packet buffer.

```csharp
using var stream = new MemoryStream();
using var packet = new BinaryWriter(stream);

packet.WriteInt32(42);
packet.WriteString("Hello World!");

byte[] buffer = packet.GetBuffer();

_user.Send(buffer); // byte[] overload
// or
_user.Send(packet); // Stream overload
```

* Remove the `LitePacket` class implementation.

> Note: The `LitePacketStream` will be kept to provide an alternative to the `BinaryReader/Writer` implementation. Note that the `LitePacketStream` uses a `BinaryReader` and `BinaryWriter` internally.

## Packet Processor

The packet processor will kinda remain the same, because it tells the receiver how to handle packet framing.

* Remove the `virtual ILitePacketStream CreatePacket(byte[] buffer)` : No need for this method since we are removing the `ILitePacketStream`. We will pass directly the raw buffer to the `HandleMessageAsync` method.
* Add a `FormatPacket(byte[])` method.
  * This method will take the raw packet buffer as parameter and allows the user to format its packet header. For example, in the default LikePacketProcessor, it would add the length of the packet content as packet header. Stored as an `int` https://github.com/Eastrall/LiteNetwork/blob/main/src/LiteNetwork/Protocol/LitePacketProcessor.cs#L12

# LiteClient / LiteServer

* Rework the `ILiteServer` abstraction to be more straightforward and simple.
* Remove the `ILiteClient` abstraction. Keeping it simple and extendable.

# Hosting

## Lite Server

* Remove `ILiteBuilder AddLiteServer<TLiteServerUser>`
* Replace `ILiteBuilder AddLiteServer<TLiteServer, TLiteServerUser>` by `ILiteBuilder AddLiteServer<TLiteServer>`
Pass directly a LiteServer implementation to the builder.
* Replace `ILiteBuilder AddLiteServer<TLiteServer, TLiteServerImplementation, TLiteServerUser>` by `ILiteBuilder AddLiteServer<TLiteServer, TLiteServerImplementation>`
If the user wants to add an extra interface to its LiteServer he can also register the interface and the implementation together in the service provider.

## Lite Client

* Remove default `ILiteBuilder AddLiteClient()` method.

# Bottom line

Thank you to @hoshinokanade for the great feedback on the library's top-level API. 😄

**Comment by Eastrall** [comment]
Changes can be found on the [`vnext`](https://github.com/Eastrall/LiteNetwork/tree/vnext) branch.

**Comment by Eastrall** [comment]
PR #48 merged.
